firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/CrashlyticsController.java

```diff
@@ -1726,76 +1732,56 @@ private void sendSessionReports(@NonNull AppSettingsData appSettings, boolean da
-      if (AnalyticsConnectorReceiver.APP_EXCEPTION_EVENT_NAME.equals(eventName)) {
-        eventLatch.countDown();
-      }
-    }
-  }
-
   /**
    * Send an App Exception event to Firebase Analytics. FA records the event asynchronously, so this
    * method returns a Task in case the caller wants to verify that the event was recorded by FA and
    * will not be lost.
    */
-  private Task<Void> recordFatalFirebaseEvent(long timestamp) {
-    final ThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
-    return Tasks.call(
-        executor,
-        new Callable<Void>() {
-          @Override
-          public Void call() throws Exception {
-            if (firebaseCrashExists()) {
-              Logger.getLogger()
-                  .d("Skipping logging Crashlytics event to Firebase, FirebaseCrash exists");
-              return null;
-            }
-            if (analyticsConnector == null) {
-              Logger.getLogger()
-                  .d("Skipping logging Crashlytics event to Firebase, no Firebase Analytics");
-              return null;
-            }
-            final BlockingCrashEventListener blockingListener = new BlockingCrashEventListener();
-            analyticsReceiver.setCrashlyticsOriginEventListener(blockingListener);
+  private void recordFatalFirebaseEvent(long timestamp) {
+    final Executor executor = backgroundWorker.getExecutor();
+    // Add another iteration of this task to the existing task chain.
+    // Keeping each iteration serialized in a chain ensures that we receive the callback for the
+    // sent event before sending another.
+    recordFatalFirebaseEventsTask =
+        recordFatalFirebaseEventsTask.continueWithTask(
+            executor,
+            (previousTask) -> {
+              if (firebaseCrashExists()) {
+                Logger.getLogger()
+                    .d("Skipping logging Crashlytics event to Firebase, FirebaseCrash exists");
+                return Tasks.forResult(null);
+              }
+              if (analyticsConnector == null) {
+                Logger.getLogger()
+                    .d("Skipping logging Crashlytics event to Firebase, no Firebase Analytics");
+                return Tasks.forResult(null);
+              }
 
-            Logger.getLogger().d("Logging Crashlytics event to Firebase");
-            final Bundle params = new Bundle();
-            params.putInt(FIREBASE_CRASH_TYPE, FIREBASE_CRASH_TYPE_FATAL);
-            params.putLong(FIREBASE_TIMESTAMP, timestamp);
+              final TaskCompletionSource<Void> source = new TaskCompletionSource<>();
 
-            analyticsConnector.logEvent(
-                FIREBASE_ANALYTICS_ORIGIN_CRASHLYTICS, FIREBASE_APPLICATION_EXCEPTION, params);
+              final CrashlyticsAppExceptionEventListener callback =
+                  () -> {
+                    // Only listen for a single event callback
+                    analyticsReceiver.setCrashlyticsAppExceptionEventListener(null);
```

Comment:
what thread is this callback triggered on? is it possible that it gets called twice before this set gets called?

Reply:
It's called on the background thread on which we receive the event back from Firebase Analytics. 2 answers to your question:

1) See this comment: https://github.com/firebase/firebase-android-sdk/pull/1433/files#diff-5e547683a03530a699d7fe85df91fa10R1743
2) If it did call more than once, it would be a no-op.